common/dbus/platform/asio_libdbus/dbus_client.cpp

```diff
@@ -135,16 +142,34 @@ error::Error DBusClient::CallMethod(
 	// We need to create a copy here because we need to make sure the handler,
 	// which might be a lambda, even with captures, will live long enough for
 	// the finished pending call to use it.
-	DBusCallReplyHandler *handler_copy = new DBusCallReplyHandler(handler);
-	if (!dbus_pending_call_set_notify(pending, HandleReply, handler_copy, FreeHandlerCopy)) {
+	auto *handler_copy = new DBusCallReplyHandler<ReplyType>(handler);
+	if (!dbus_pending_call_set_notify(
+			pending, HandleReply<ReplyType>, handler_copy, FreeHandlerCopy<ReplyType>)) {
 		return MakeError(MessageError, "Failed to set reply handler");
 	}
 
 	return error::NoError;
 }
 
+template error::Error DBusClient::CallMethod(
+	const string &destination,
+	const string &path,
+	const string &iface,
+	const string &method,
+	DBusCallReplyHandler<expected::ExpectedString> handler);
+
+template <>
+void DBusClient::AddSignalHandler(
```

Comment:
Ended up using both template instantiation *and* template specialization! Advanced stuff! :man_scientist: :smile: 

But I agree that this combo solves the problem well.

Reply:
> Ended up using both template instantiation _and_ template specialization! Advanced stuff! 👨‍🔬 😄

No idea... This is just what I arrived to by trying various combinations of things. :joy: